beacon-chain/core/epoch/epoch_processing.go

```diff
@@ -191,3 +192,86 @@ func ProcessEjections(state *pb.BeaconState) (*pb.BeaconState, error) {
 	}
 	return state, nil
 }
+
+// ProcessValidatorRegistry computes and sets new validator registry fields,
+// reshuffles shard committees and returns the recomputed state.
+//
+// Spec pseudocode definition:
+//	Set state.validator_registry_latest_change_slot = state.slot.
+//	Set state.shard_committees_at_slots[:EPOCH_LENGTH] = state.shard_committees_at_slots[EPOCH_LENGTH:].
+//	Set state.shard_committees_at_slots[EPOCH_LENGTH:] =
+//  	get_new_shuffling(state.latest_randao_mixes[(state.slot - EPOCH_LENGTH) %
+//  	LATEST_RANDAO_MIXES_LENGTH], state.validator_registry, next_start_shard)
+//  	where next_start_shard = (state.shard_committees_at_slots[-1][-1].shard + 1) % SHARD_COUNT.
+func ProcessValidatorRegistry(
+	state *pb.BeaconState) (*pb.BeaconState, error) {
+
+	epochLength := int(params.BeaconConfig().EpochLength)
+	randaoMixesLength := params.BeaconConfig().LatestRandaoMixesLength
+	shardCount := params.BeaconConfig().ShardCount
+	lastShardAndCommitteesAtSlots := len(state.ShardAndCommitteesAtSlots) - 1
+	lastCommiteeFromLastSlot := len(state.ShardAndCommitteesAtSlots[lastShardAndCommitteesAtSlots].ArrayShardAndCommittee) - 1
+	nextStartShard := (state.ShardAndCommitteesAtSlots[lastShardAndCommitteesAtSlots].ArrayShardAndCommittee[lastCommiteeFromLastSlot].Shard +
+		1) % shardCount
+
+	var randaoHash32 [32]byte
+	copy(randaoHash32[:], state.LatestRandaoMixesHash32S[(state.Slot-uint64(epochLength))%randaoMixesLength])
+
+	state.ValidatorRegistryLastChangeSlot = state.Slot
+	for i := 0; i < epochLength; i++ {
+		state.ShardAndCommitteesAtSlots[epochLength+i] = state.ShardAndCommitteesAtSlots[i]
+	}
+	newShuffledCommittees, err := validators.ShuffleValidatorRegistryToCommittees(
+		randaoHash32,
+		state.ValidatorRegistry,
+		nextStartShard,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not shuffle validator registry for commtitees: %v", err)
+	}
+
+	for i := 0; i < epochLength; i++ {
+		state.ShardAndCommitteesAtSlots[i] = newShuffledCommittees[i]
+	}
+	return state, nil
+}
+
+// ProcessPartialValidatorRegistry processes the portion of validator registry
+// fields, it doesn't set registry latest change slot. This only gets called if
+// validator registry update did not happen.
+//
+// Spec pseudocode definition:
+//	Set state.shard_committees_at_slots[:EPOCH_LENGTH] = state.shard_committees_at_slots[EPOCH_LENGTH:].
+//  Let epochs_since_last_registry_change = (state.slot - state.validator_registry_latest_change_slot) // EPOCH_LENGTH.
```

Comment:
Did you check this with a local godocs to see if it looks OK?

Reply:
Checked and verified it looks good 👍